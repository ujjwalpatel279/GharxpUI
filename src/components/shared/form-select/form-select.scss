// ========================================================================
// COMPONENTS/SHARED/FORM-SELECT
// ========================================================================

@import '~sassHelpers';

.c-form-select-wrapper {
  position: relative;

  &::before {
    content: '';
    position: absolute;
    right: spacer(2);
    top: 50%;
    width: 0px;
    height: 0px;
    transform: translateY(-50%);
    border-style: solid;
    border-width: 7px 6.5px 0px;
    border-color: color-brand(primary) transparent transparent;
  }
}

.c-form-select {
  height: spacer(10);
  width: 100%;
  padding-left: spacer(2);
  padding-right: spacer(4);
  border: 1px solid color(grey3);
  border-radius: $border-radius;
  background-color: white;
  appearance: none;

  &::-ms-expand {
    display: none;
  }

  &.isRequired {
    border-left: 5px solid color(navy1);
  }

  &.isInvalid {
    background-color: color(errorLight);
    border-color: color(errorDark);
  }

  &:hover,
  &:focus {
    border-color: color(grey2);
  }
}
