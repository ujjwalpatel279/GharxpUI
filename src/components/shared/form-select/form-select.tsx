import React, { forwardRef, ChangeEvent } from 'react';

interface OptionsInterface {
  id: number | string;
  value: string;
  displayText?: string;
  selectable?: boolean;
}

interface FormFieldProps {
  id: string;
  options: OptionsInterface[];
  error?: boolean;
  displayDefaultSelect?: boolean;
  changeHandler?: (e: ChangeEvent<HTMLSelectElement>) => void;
  value?: number;
}

import './form-select.scss';

export const FormSelect = forwardRef(
  (
    { id, options, error, displayDefaultSelect, changeHandler, value }: FormFieldProps,
    ref: React.Ref<HTMLSelectElement>,
  ) => {
    let selectClasses = 'c-form-select';

    if (error) {
      selectClasses += ' isInvalid';
    }

    return (
      <div className="c-form-select-wrapper">
        <select
          name={id}
          id={id}
          className={selectClasses}
          ref={ref}
          aria-invalid={error ? 'true' : 'false'}
          data-testid="form-select"
          onChange={changeHandler}
          value={value}
        >
          {(displayDefaultSelect ?? true) && <option value="">Please select</option>}
          {options.map((option: OptionsInterface) => (
            <option key={option.id} value={option.value} disabled={option.selectable ?? true ? false : true}>
              {option.displayText ?? option.value}
            </option>
          ))}
        </select>
      </div>
    );
  },
);

FormSelect.displayName = 'FormSelect';
