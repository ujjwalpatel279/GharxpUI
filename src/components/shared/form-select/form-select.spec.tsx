import React from 'react';
import { render, cleanup } from '@testing-library/react';

import { FormSelect } from './form-select';

const mockOptions = [
  {
    id: 1,
    value: 'One',
  },
  {
    id: 2,
    value: 'Two',
  },
];

describe('FormInput Component', () => {
  afterEach(cleanup);

  it('renders the form select component', () => {
    const { container } = render(<FormSelect id="testSelect" options={mockOptions} ref={() => null} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('aria-invalid is false', () => {
    const { getByTestId } = render(<FormSelect id="testSelect" options={mockOptions} error={false} ref={() => null} />);

    expect(getByTestId('form-select').getAttribute('aria-invalid')).toBe('false');
  });

  it('aria-invalid is true', () => {
    const { getByTestId } = render(<FormSelect id="testSelect" options={mockOptions} error={true} ref={() => null} />);

    expect(getByTestId('form-select').getAttribute('aria-invalid')).toBe('true');
  });

  it('first option is "Please select"', () => {
    const { getByTestId } = render(<FormSelect id="testSelect" options={mockOptions} ref={() => null} />);

    const select = getByTestId('form-select') as HTMLInputElement;
    expect(select.value).toBe('');
  });
});
