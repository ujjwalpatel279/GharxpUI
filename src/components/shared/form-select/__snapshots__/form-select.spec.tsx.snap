// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Component renders the form select component 1`] = `
<div
  class="c-form-select-wrapper"
>
  <select
    aria-invalid="false"
    class="c-form-select"
    data-testid="form-select"
    id="testSelect"
    name="testSelect"
  >
    <option
      value=""
    >
      Please select
    </option>
    <option
      value="One"
    >
      One
    </option>
    <option
      value="Two"
    >
      Two
    </option>
  </select>
</div>
`;
